### PR TITLE
13034 error messages: mismatched labels

### LINKED ISSUE #13034: Improve error message for optional arguments mismatches
I spent little but somehow too much time alpha renaming these expressions to see where the problem could be before realizing it was simply an optional vs non optional issue (maybe I'm tired). 

Could maybe the error message be improved to directly mention optionability differences ? 
 
```ocaml
       Values do not match:
         val v3 :
           ?fold:('a, 'b) Vec.fold ->
           ?make:('b -> 'b -> 'b -> 'a) -> 'b t -> 'a t
       is not included in
         val v3 :
           fold:('v3, 'a) Vec.fold ->
           make:('a -> 'a -> 'a -> 'v3) -> 'a t -> 'v3 t
       The type
         ?fold:('a, 'b) Vec.fold ->
         ?make:('b -> 'b -> 'b -> 'a) -> 'b t -> 'a t
       is not compatible with the type
         fold:('c, 'd) Vec.fold ->
         make:('d -> 'd -> 'd -> 'c) -> 'd t -> 'c t
```
(OCaml v4.14.1)

### PR DESCRIPTION
**Pull request by Octachron**
This PR adds a specialized explanation for mismatched function labels in type error messages.
For instance, in

```ocaml
module Label1 : sig
  val f: int -> unit
end = struct
  let f ~x = ()
end
```
the existing error message
>```Error: Signature mismatch:
>       Modules do not match:
>         sig val f : x:'a -> unit end
>       is not included in
>         sig val f : int -> unit end
>       Values do not match:
>         val f : x:'a -> unit
>       is not included in
>         val f : int -> unit
>       The type "x:'a -> unit" is not compatible with the type "int -> unit"
is extended with
>```
>       The argument "x" is labeled, but an unlabeled argument was expected
>```

This complements the existing specialized error message for syntactic argument with a label incompatible with the expected type.

close  #13034

### CODE REVIEW

File: typing/ctype.ml [hunk 1]
```diff
@@ -2835,15 +2835,17 @@ and unify3 uenv t1 t1' t2 t2' =
...
+          if
+            compatible_labels ~in_pattern_mode:(in_pattern_mode uenv) l1 l2
+          then begin unify uenv t1 t2; unify uenv u1 u2;
+            match is_commu_ok c1, is_commu_ok c2 with
+            | false, true -> set_commu_ok c1
+            | true, false -> set_commu_ok c2
+            | false, false -> link_commu ~inside:c1 c2
+            | true, true -> ()
+          end else
+            raise_for Unify (Function_label_mismatch {got=l1; expected=l2})
```
Comment (gasche) [inline typing/ctype.ml]: Nitpick: wouldn't this be easier on the eye with the error case first?

```ocaml
if not (compatible_labels ...) then
  raise_for ...;
unify uenv t1 t2; unify uenv u1 u2;
begin match ... with
  ...
end
```

(Same question in the other, simpler cases.)

Comment (Octachron) [reply typing/ctype.ml]: Why not?

Comment (samsa1) [reply typing/ctype.ml]: Woudn't it be even simpler if we moved the raise inside the `compatible_labels` function ?
This would reduce code duplication (and even more if more label comparison are added).

Comment (Octachron) [reply typing/ctype.ml]: There are calls to `compatible_labels` in `mcomp` and the subtyping mode that doesn't raise the exception, either because we are only checking compatibility or because we are storing an error trace for later use.

However, we could have a `raise_on_incompatible_labels` function which has the advantage of keeping the control flow obvious.

File: testsuite/tests/typing-modules/inclusion_errors.ml [hunk 2]
```diff
@@ -1853,3 +1853,183 @@ Error: Signature mismatch:
...
+       Modules do not match:
+         sig type t = x:int -> int end
+       is not included in
+         sig type t = int -> int end
+       Type declarations do not match:
+         type t = x:int -> int
+       is not included in
+         type t = int -> int
+       The type "x:int -> int" is not equal to the type "int -> int"
+       The argument "x" is labeled, but an unlabeled argument was expected
```
Comment (gasche) [inline testsuite/tests/typing-modules/inclusion_errors.ml]: Proposal: `The argument has label "x", but ...`.

File: testsuite/tests/typing-modules/inclusion_errors.ml [hunk 3]
```diff
@@ -1853,3 +1853,183 @@ Error: Signature mismatch:
...
+       Modules do not match:
+         sig val f : x:'a -> unit end
+       is not included in
+         sig val f : int -> unit end
+       Values do not match:
+         val f : x:'a -> unit
+       is not included in
+         val f : int -> unit
+       The type "x:'a -> unit" is not compatible with the type "int -> unit"
+       The argument "x" is labeled, but an unlabeled argument was expected
```
Comment (gasche) [inline testsuite/tests/typing-modules/inclusion_errors.ml]: `The argument has label "x"`, ...`

File: testsuite/tests/typing-modules/inclusion_errors.ml [hunk 4]
```diff
@@ -1853,3 +1853,183 @@ Error: Signature mismatch:
...
+       Modules do not match:
+         sig val f : ?x:'a -> unit end
+       is not included in
+         sig val f : int -> unit end
+       Values do not match:
+         val f : ?x:'a -> unit
+       is not included in
+         val f : int -> unit
+       The type "?x:'a -> unit" is not compatible with the type "int -> unit"
+       The argument "?x" is optional, but an unlabeled argument was expected
```
Comment (gasche) [inline testsuite/tests/typing-modules/inclusion_errors.ml]: `The argument has label `?x`, ...`.

File: testsuite/tests/typing-modules/inclusion_errors.ml [hunk 5]
```diff
@@ -1853,3 +1853,183 @@ Error: Signature mismatch:
...
+       Modules do not match:
+         sig val f : ?x:'a -> unit end
+       is not included in
+         sig val f : x:int -> unit end
+       Values do not match:
+         val f : ?x:'a -> unit
+       is not included in
+         val f : x:int -> unit
+       The type "?x:'a -> unit" is not compatible with the type "x:int -> unit"
+       The optional argument "?x" was not expected to be optional
```
Comment (gasche) [inline testsuite/tests/typing-modules/inclusion_errors.ml]: `Labels "?x" and "x" do not match.`

File: testsuite/tests/typing-modules/inclusion_errors.ml [hunk 6]
```diff
@@ -1853,3 +1853,183 @@ Error: Signature mismatch:
...
+       Modules do not match:
+         sig val f : 'a -> unit end
+       is not included in
+         sig val f : ?x:int -> unit end
+       Values do not match:
+         val f : 'a -> unit
+       is not included in
+         val f : ?x:int -> unit
+       The type "'a -> unit" is not compatible with the type "?x:int -> unit"
+       An optional argument "?x" was expected
```
Comment (gasche) [inline testsuite/tests/typing-modules/inclusion_errors.ml]: `A label "?x" was expected.`

**Review by gasche — COMMENTED** [review]
I am uneasy with the wording `The optional argument "?x"`, because `?x` is a label and not an argument`. I proposed some reformulations that avoid this issue. Also in the case of a mismatch between an optional and a non-optional label, I propose to simplify with a more approximative but shorter error message. Feel free to adapt the suggestions or discard them outright.

File: testsuite/tests/typing-modules/inclusion_errors.ml (hunk 2 above)
Comment (Octachron) [reply testsuite/tests/typing-modules/inclusion_errors.ml]: Counterproposal: `The first argument is labeled "x"`

File: testsuite/tests/typing-modules/inclusion_errors.ml (hunk 5 above)
Comment (Octachron) [reply testsuite/tests/typing-modules/inclusion_errors.ml]: I disagree with repeating twice the same name with one minute change.
I propose: `Label "?x" was expected to not be optional`.

Comment (gasche) [reply testsuite/tests/typing-modules/inclusion_errors.ml]: Wait, what would the error be on a mismatch between `?x:int -> unit` and `y:int -> unit`? I assumed that the use of `x` on both sides was irrelevant to the current situation and wording.

File: testsuite/tests/typing-modules/inclusion_errors.ml (hunk 2 above)
Comment (gasche) [reply testsuite/tests/typing-modules/inclusion_errors.ml]: Fine with me.

File: testsuite/tests/typing-modules/inclusion_errors.ml (hunk 5 above)
Comment (Octachron) [reply testsuite/tests/typing-modules/inclusion_errors.ml]: In this case, `Labels "?x" and "y" do not match` seems a good simplification to me (the previous error message was `The optional argument "?x" was expected to be a labeled argument named "y"`)

File: testsuite/tests/typing-misc/labels.ml [hunk 7]
```diff
@@ -172,7 +172,8 @@ Line 1, characters 26-33:
                               ^^^^^^^
 Error: This expression has type "x:'a -> unit"
        but an expression was expected of type "unit -> unit"
-       The argument "x" is labeled, but an unlabeled argument was expected
+       The first argument is labeled "x",
+       but an unlabeled argument was expected
```
Comment (gasche) [inline testsuite/tests/typing-misc/labels.ml:176]: Side remark: the absence of a final `.` at the end of these captialized sentences is odd, but this is orthogonal to the current change.

**Review by gasche — APPROVED** [review]

File: testsuite/tests/typing-misc/labels.ml (hunk 7 above)
Comment (Octachron) [reply testsuite/tests/typing-misc/labels.ml:176]: I agree and generally struggle to remember to not add period in those explanations.
